### DISCUSSION TITLE
Getting Typeclass Name of Object

### DISCUSSION THREAD

**Discussion by mOctave** [evennia/evennia#3158] [Evennia Questions] [answered]
I'm not sure if this is a bug with my code or Evennia, but I'm guessing it's user error so I'm posting it here.

I'm trying to list everything in a room along with properties, but I want to avoid certain kinds of thing. For this reason I thought to use `__class__` and `__name__` as follows:

```
		for obj in room.contents:
			if not obj.__class__.__name__ in ["Character", "Exit", "Currency"]:
				...
```

~~However, upon triggering the function, `__class__` doesn't work, which I'm guessing is because of how typeclasses are set up.~~

Two questions:
~~1. Is this just user error, or is there a bug here with Evennia as well?~~
2. How am I supposed to go about doing this kind of thing? Presumably there's a common method that isn't what I've just done.

_Edit:_ It looks like that might have been triggered by something else, not that part of the code.

**Comment by jaborsh** [comment]
The best answer probably requires an understanding of what certain things you're trying to omit. Using your array as the assumption, I'd suggest you use Typeclasses. All items in Evennia can be identified by their senior-most parent. For characters, the Typeclass is probably Character. Likewise with the others. Take a look at this: https://www.evennia.com/docs/latest/Components/Typeclasses.html?highlight=typeclass

As for code syntax, you can then rewrite your if statement to look for inheritance.

**Comment by jaborsh** [comment]
As a side note, the `return_appearance` hook for rooms already does something similar to separate exits from other contents found in a room. You could review the code used there and adapt it to your preferences.

**Comment by Griatch** [comment, marked as answer]
Typeclasses does have `__class__`, but you already found out the error about that :) 

If you are using Typeclasses to separate your types of data, you should normally do so because each typeclass contains unique Python code. In this case, there is no need to cook your own solution to check inheritance - use [evennia.utils.utils.inherits_from](https://www.evennia.com/docs/latest/api/evennia.utils.utils.html?highlight=inherits_from#evennia.utils.utils.inherits_from) - this will genereally handle inheritance in a way that makes sense. This probably makes sense for your `Character`, `Exit` and `Currency` example, since they are all likely very different entities, with different code functionality. 

If your entities are sharing code but you just want to organize them into different groupings, another approach is to [Tag](https://www.evennia.com/docs/latest/Components/Tags.html) them and use the tag to identify what 'type' of entity it is. This could be used to group, say, martial guilds and magical guilds separately - the code to represent a guild (normally in a [Script](https://www.evennia.com/docs/latest/Components/Scripts.html)) may be the same from guild to guild, tagging makes a quick way to add meta-information to group them. Zoning is another common use for Tags.
